Optimize extension auto-reloading

paescuj opened the pull request:
## Scope

What's changed:

- Watch local package extensions (#20380)

- Debounce watcher events
  - Prevents consecutive reloads when rebuilding bundle or hybrid extensions, which results in two `change` events about the same time (for `api.js` and `app.js` files)
  - Prevents consecutive reloads when building a bundle or hybrid extensions for the first time (`add` event)
  - Reduces consecutive reloads when rebuilding multiple extensions at the same time, like in #20440
  - Prevents error message about missing `dist` files during removal of an extension (thus the `2s` delay for `delete` events)
  - The `750ms` delay for `change` seems good enough to catch most cases and at the same time still lead to what feels like an almost instant reload

     https://github.com/directus/directus/assets/5363448/f48a7dc3-3127-4dfe-a73d-74378e41db3b
  - See also https://github.com/directus/directus/pull/20849#issuecomment-1895636360 for additional insights
- Respect the `PACKAGE_FILE_LOCATION` config option for watching the root `package.json` file
- Introduced fixes for `chokidar` watching unwanted files

## Potential Risks / Drawbacks

Tiny delays, but this should be made up for by the prevented reloads

## Review Notes / Questions

Tested with extension installed as dependency as well as direct and linked (`npm run link`) package extensions in `extensions` folder on
✅ macOS
✅ Linux (Ubuntu)
✅ Windows
✅ Docker (macOS)

- You may want to log `this.watcher.getWatched()` after updates to see the list of currently watched files
- You may want to log `this.reloadQueue.size` to see if and when it takes effect
- Over the long-term, I'd like to move away from `chokidar` to get rid of the applied workarounds. Vite and others are considering migration to [parcel/watcher](https://github.com/parcel-bundler/watcher), but it's currently lacking a few features... 

---

Fixes #20380
Fixes #20440
In addition to #20849

paescuj committed:
Debounce extension watcher events & watch local package extension

paescuj committed:
Add changeset

br41nslug commented on api/src/extensions/manager.ts:
```diff
@@ -307,10 +308,10 @@ export class ExtensionManager {
 		logger.info('Watching extensions for changes...');
 
-		const extensionDirUrl = pathToRelativeUrl(getExtensionsPath());
+		const extensionsDir = path.posix.resolve(getExtensionsPath());
```
Do we need `path.posix.resolve` here? This function is known to have some footguns resulting in non-posix compliant paths on Windows in certain situations (for example `path.posix.resolve(process.cwd(), 'foo') -> '/Users/Brainslug/C:\\Users\\Brainslug/foo'`)

ComfortablyCoding replied:
The usage of posix (213 and 236) could be causing the issue in https://github.com/directus/directus/issues/18721. We should check if this is resolved once paths are made OS agnostic.

paescuj replied:
Nope, that's https://github.com/paulmillr/chokidar/issues/1051

ComfortablyCoding replied:
It might not be as that one reports it as a macOS issue while #18721 reports that it works on macOS/linux but not windows.

paescuj replied:
Hmm, I cannot imagine how that could affect the Docker container, as the OS will always be Linux.

But as outlined in https://github.com/directus/directus/pull/21111#discussion_r1455898895, happy to switch to OS agnostic `path.resolve`.
@nickrum Do you remember the reason for using `path.posix`?

nickrum replied:
Yup, I just answered [here](https://github.com/directus/directus/pull/21111#discussion_r1456064371) before I saw your comment 😅

br41nslug commented on api/src/extensions/manager.ts:
```diff
@@ -307,10 +308,10 @@ export class ExtensionManager {
 
-		const localExtensionUrls = NESTED_EXTENSION_TYPES.flatMap((type) => {
-			const typeDir = path.posix.join(extensionDirUrl, pluralize(type));
+		const nestedExtensions = NESTED_EXTENSION_TYPES.flatMap((type) => {
+			const typeDir = path.posix.join(extensionsDir, pluralize(type));
```
This line is creating invalid paths resulting in legacy extension folders not being watched consistently on Windows.

![image](https://github.com/directus/directus/assets/9389634/e3aae01e-87d3-4345-a371-1c30afd5fe4f)

paescuj replied:
<sub>Also in response to https://github.com/directus/directus/pull/21111/files#r1455398161</sub>

I've tested it on all platforms, including Windows.
It works on my end, the legacy extensions are watched too.
The paths are passed to `chokidar` in which they are correctly interpreted.

I went with `path.posix` on line 311 to be in accordance with `path.posix` on line 313 which was already there before. 
I'm not sure about the original reason for using `path.posix`, an assumption could be that there were concerns about globs with Windows-path-style.
I didn't want to change that as part of this PR, because as mentioned I don't know about the reason and it worked fine in my tests.

However, happy to update this if we can get some more clarity 👍

nickrum replied:
The chokidar docs specifically mention that "globs must not contain windows separators (\), because that's how they work by the standard". That's why on line 311 `pathToRelativeUrl` was used, which converts any path to a relative URL (making it a relative path first ensures that we do not have to deal with windows vs posix root specifiers). So, even if it might works, maybe we should keep it that way to adhere to the standard 🤔

paescuj replied:
Thank you @nickrum ❤️ 🙏 

The thing is that I need the absolute path to be able to apply https://github.com/directus/directus/pull/21111/files#diff-09fef1df59269b49b62b7e749e695089e0e51baa3f666521c50a48a58a7820f6R332

So in that case, could we use the following?
```js
const extensionsDir = pathToRelativeUrl(path.resolve(getExtensionsPath()));
```

paescuj replied:
Alright! `pathToRelativeUrl` obviously won't work for absolute paths, so I gave it a try and removed all usages of `posix`. Turns out that works perfectly fine 👍🥳 
I think that is either because Node.js' `path` API already escapes the paths correctly or because `chokidar`'s statement is simply outdated.

The generated paths:
<img width="900" src="https://github.com/directus/directus/assets/5363448/9d5015fb-c701-4e12-9c70-6cba70278eb7">
 
Excerpt of chokidar's watched paths:
<img width="900" src="https://github.com/directus/directus/assets/5363448/ece410da-dc87-4c1d-9d08-35a1b8b66908">

Thanks @br41nslug thanks for checking on that, and thanks again @nickrum for the info! 🚀

nickrum replied:
> The thing is that I need the absolute path to be able to apply https://github.com/directus/directus/pull/21111/files#diff-09fef1df59269b49b62b7e749e695089e0e51baa3f666521c50a48a58a7820f6R332

Do we actually need absolute paths for this to work, though? Couldn't this also be converted to a relative URL? Wouldn't that make sense given it is a glob pattern as well? I'm a little bit worried using something which is explicitly advised against. It might only work by chance and could break in a future minor/patch update 🤔

paescuj replied:
Two things before I spend more time on this:
1. `chokidar` is more or less in unmaintained state, so I doubt there will be updates which break that behavior (or updates at all)
2. Over the long-term, I'd like to move away from `chokidar` anyway (see OP) because of 1) and the quirks it comes with

nickrum replied:
Ah, I didn't even know that. Are there any good alternatives?

I really don't want to make your live any harder 🫣 So, if it works and as long as we keep in mind that this could break once we move away from chokidar, I'd be fine with it.

paescuj replied:
> Ah, I didn't even know that. Are there any good alternatives?

Unfortunately, not really as of now 😢 There's [`parcel/watcher`](https://github.com/parcel-bundler/watcher) which looks promising but it's missing a high-level API, similar to `fs.watch`. I think https://github.com/vitejs/vite/issues/13593#issuecomment-1776906216 sums it up pretty well, for our needs too.

> I really don't want to make your live any harder 🫣

All good! ❤️ I just wanted to have that mentioned as basis for further discussion. I didn't mean to say that we can't make any further adjustments.

> So, if it works and as long as we keep in mind that this could break once we move away from chokidar, I'd be fine with it.

Yeah, that's what I'm also leaning towards at the moment.

br41nslug replied:
Seems to work fine, in my tests today 🤔 not sure if due to the changes or due to flawed testing in my original review.

br41nslug reviewed (commented):
Loving the fixes 🚀😄 However the Windows path issues do worry me 😬

br41nslug commented:
I know we are not currently doing this anyway but should we include the 2 forgotten children of extensions in the auto-reloading? Migrations and Email Templates, specifically the email templates could be very valuable to have auto reloaded while working on them.

paescuj committed:
Remove leftover from testing

paescuj commented on api/src/extensions/manager.ts:
```diff
@@ -322,17 +323,30 @@ export class ExtensionManager {
 		this.watcher
-			.on('add', () => this.reload())
-			.on('change', () => this.reload())
-			.on('unlink', () => this.reload());
+			.on(
+				'add',
+				debounce(() => this.reload(), 500),
+			)
+			.on(
+				'change',
+				debounce(() => this.reload(), 650),
+			)
+			.on(
+				'unlink',
+				debounce(() => this.reload(), 2000),
+			);
```
I'd rather not make this configurable. I don't see a real reason for this.
The debouncing is intended as an optimization, with values that will cover most cases.
However, we can never cover all cases (for example edge-cases like in #20440).
Even if we make it configurable, it doesn't really make sense to set the values higher, because then it takes too long for single file changes.
On the other side, IMO the values are small enough not to impair everyday use.

paescuj committed:
Use OS agnostic paths

ComfortablyCoding committed:
Merge branch 'main' into extensions-reload

ComfortablyCoding commented on .changeset/witty-cobras-add.md:
```diff
@@ -0,0 +1,5 @@
+'@directus/api': patch
+---
+
+Optimized extension auto-reload behaviour and added auto-reload support for local package extensions
```
```suggestion
Optimized extension auto-reload behaviour and fixed auto-reload for local package extensions
```

ComfortablyCoding reviewed (commented):
LGTM!

Waiting on confirmation from @br41nslug that all is well on a windows native machine before approving.

ComfortablyCoding commented:
Two DX non blocking notes
- We should add some sort of log for the code change case. ATM it shows "Reloading extensions" with nothing to indicate that the reload was completed. Video attached
- When testing it was strange when I did not see the reload message after a change was made. It made me think the watcher was not working. Potentially adding the message mentioned above might resolve this. 

https://github.com/directus/directus/assets/44623501/56b342e4-7e24-474e-a19c-34400979fc86

br41nslug reviewed (approved):
While theoretically possible to lose the last extension update done (if done within the 500ms debounce of the last update) that is quite fast and definitely an edge case 😄

paescuj committed:
Log info after extensions reload

The idea is to have both "Reloading extensions" and a subsequent
"Extensions reloaded" message, but since they appear at the same time,
we can just omit the former.
The issue with just having the "Reloading extensions" message is that
it's not clear whether the reloading was actually successful.

paescuj committed:
Update changeset

Co-authored-by: daedalus <44623501+ComfortablyCoding@users.noreply.github.com>
